File: code/Java/state/gumballmachine/NoQuarterState.java
package state.gumballmachine;

public class NoQuarterState implements State{
    private GumballMachine gumballMachine;

    public NoQuarterState(GumballMachine gm) {
        this.gumballMachine = gm;
    }

    @Override
    public void insertQuarter() {
        System.out.println("You insert a quarter");
        gumballMachine.setCurState(gumballMachine.getHasQuarterState());
    }

    @Override
    public void ejectQuarter() {
        System.out.println("You haven't insert a quarter");
    }

    @Override
    public void turnCrank() {
        System.out.println("You turned but there is no quarter");
    }

    @Override
    public void dispense() {
        System.out.println("You need to pay first");
    }

    @Override
    public void refill() {    }

//    @Override
//    public String toString() {
//        return "NoQuarterState{" +
//                "gumballMachine=" + this.gumballMachine +
//                '}';
//    }
}

File: code/C++/state/noquarterstate.cpp
#include "noquarterstate.h"

NoQuarterState::NoQuarterState(GumballMachine * const gm):gumballMachine(gm) {}

void NoQuarterState::insertQuarter()
{
	std::cout << "You insert a quarter" << std::endl;
	this->gumballMachine->setState(this->gumballMachine->getHasQuarterState());
}

void NoQuarterState::ejectQuarter()
{
	std::cout << "You haven't insert a quarter" << std::endl;
}

void NoQuarterState::turnCrank()
{
	std::cout << "You turned but there is no quarter" << std::endl;
}

void NoQuarterState::dispense()
{
	std::cout << "You need to pay first" << std::endl;
}

void NoQuarterState::refill()
{

}

void NoQuarterState::display()
{
	std::cout << "NoQuarterState, count:" << this->gumballMachine->getCount() << std::endl;
}